### ISSUE TITLE
Parallelizing step3

### ISSUE THREAD

**Issue by matbesancon** [Gnimuc/Hungarian.jl#17]
We're using Hungarian.jl in the https://github.com/ZIB-IOL/FrankWolfe.jl package to solve LPs over the Birkhoff polytope. 

It scales well in terms of memory footprint but badly in terms of time, mostly because of `step3!`, where most of the time seems to be spent in this block:
```julia
    @inbounds for j in colUncoveredIdx, i in rowUncoveredIdx
        cost = costMat[i, j] + Δcol[j] + Δrow[i]
        if cost <= h
            if cost != h
                h = cost
                empty!(minLocations)
            end
            push!(minLocations, (i, j))
        end
    end
```

this would be non-trivial to parallelize because of the shared bound `h` and array `minLocations` but could be worth it.

**Comment by Gnimuc** [comment]
> It scales well in terms of memory footprint but badly in terms of time, mostly because of `step3!`, where most of the time seems to be spent in this block:

I think the hot-pot depends on the problem size. Yes, for small matrices, `step3` is the bottleneck, but I'm not sure how much performance gain we can benefit from multi-threading.

**Comment by Gnimuc** [comment]
Looks like I tried something similar years ago: 
https://github.com/Gnimuc/Hungarian.jl/blob/a60d30352eb4ec02ffe4dae1515805de49ce09b9/src/Munkres.jl#L343

**Comment by matbesancon** [comment]
Do you get other crititcal points when matrix sizes grow larger? This was for a 1000x1000 dense matrix

**Comment by Gnimuc** [comment]
https://github.com/Gnimuc/Hungarian.jl/blob/fe5c4024eadeb250cf82512b69f52b210883fae2/src/Munkres.jl#L91-L108

This marking step is critical for large matrices like 4000x4000 or 8000x8000. You can find some benchmark reports in https://github.com/Gnimuc/Hungarian.jl/pull/15.

**Comment by Gnimuc** [comment]
What's the problem size of your use case?

**Comment by matbesancon** [comment]
the problem go up to `npixels * npixels` for image problems so as big as you want, `150528 * 150528` in some cases but we couldn't run Hungarian.jl for these sizes

**Comment by Gnimuc** [comment]
That's really large. Hungarian is an O(n^3) algorithm and may not suite for this kinda problem.

Would you like to make a PR with above-mentioned multi-threading support? I'm happy to merge if the benchmark reports are green. 

BTW, this is released under public domain, so you could also make the code tailored to your use case and embed it in your package.

**Comment by matbesancon** [comment]
> That's really large. Hungarian is an O(n^3) algorithm and may not suite for this kinda problem.

Yes this is the kind of extreme limit we'd like to solve, it's okay to have the method work for smaller sizes of course.

I'll probably try to reimplement the multithreaded version and compare the two + a LP solver
